File: packages/core-ts/src/storage/config-file.ts
import { existsSync } from 'node:fs'
import { mkdir, readFile, unlink, writeFile } from 'node:fs/promises'
import { dirname, join } from 'node:path'
import envPaths from 'env-paths'
import {
  decryptData,
  deriveFileKey,
  derivePassphraseKey,
  encryptData,
  LEGACY_PBKDF2_ITERATIONS,
  PBKDF2_ITERATIONS
} from './encryption.js'
import { getMachineId, getUsername } from './machine-id.js'

const paths = envPaths('mcp', { suffix: '' })
const DEFAULT_CONFIG_PATH = join(paths.config, 'config.enc')

interface ConfigStore {
  version: 1
  servers: Record<string, Record<string, string>>
}

// Allow overriding config path for testing
let configPathOverride: string | null = null

export function setConfigPath(path: string | null): void {
  configPathOverride = path
}

function getConfigPath(): string {
  return configPathOverride ?? DEFAULT_CONFIG_PATH
}

const MAX_RETRIES = 3
const BASE_DELAY_MS = 100

async function withRetry<T>(fn: () => Promise<T>): Promise<T> {
  for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
    try {
      return await fn()
    } catch (err: unknown) {
      const isLocked = err instanceof Error && 'code' in err && (err as NodeJS.ErrnoException).code === 'EBUSY'
      if (!isLocked || attempt === MAX_RETRIES - 1) throw err
      await new Promise((r) => setTimeout(r, BASE_DELAY_MS * 2 ** attempt))
    }
  }
  throw new Error('Unreachable')
}

async function getKey(): Promise<CryptoKey> {
  const [machineId, username] = await Promise.all([getMachineId(), getUsername()])
  return deriveFileKey(machineId, username)
}

async function loadStore(): Promise<ConfigStore> {
  const configPath = getConfigPath()
  if (!existsSync(configPath)) {
    return { version: 1, servers: {} }
  }

  const [machineId, username] = await Promise.all([getMachineId(), getUsername()])
  const data = await readFile(configPath)

  try {
    // Try with current iterations
    const key = await deriveFileKey(machineId, username, PBKDF2_ITERATIONS)
    const json = await decryptData(key, data)
    return JSON.parse(json) as ConfigStore
  } catch (err) {
    try {
      // Fallback to legacy iterations
      const legacyKey = await deriveFileKey(machineId, username, LEGACY_PBKDF2_ITERATIONS)
      const json = await decryptData(legacyKey, data)
      const store = JSON.parse(json) as ConfigStore
      // Auto-migrate to current iterations
      await saveStore(store)
      return store
    } catch {
      // If legacy also fails, throw the original error
      throw err
    }
  }
}

async function saveStore(store: ConfigStore): Promise<void> {
  const configPath = getConfigPath()
  const dir = dirname(configPath)
  if (!existsSync(dir)) {
    await mkdir(dir, { recursive: true })
  }
  const key = await getKey()
  const encrypted = await encryptData(key, JSON.stringify(store))
  await withRetry(() => writeFile(configPath, encrypted))
}

export async function readConfig(serverName: string): Promise<Record<string, string> | null> {
  const store = await loadStore()
  return store.servers[serverName] ?? null
}

export async function writeConfig(serverName: string, config: Record<string, string>): Promise<void> {
  const store = await loadStore()
  store.servers[serverName] = config
  await saveStore(store)
}

export async function deleteConfig(serverName: string): Promise<void> {
  const store = await loadStore()
  delete store.servers[serverName]

  const configPath = getConfigPath()
  if (Object.keys(store.servers).length === 0) {
    if (existsSync(configPath)) {
      await unlink(configPath)
    }
  } else {
    await saveStore(store)
  }
}

export async function listConfigs(): Promise<string[]> {
  const store = await loadStore()
  return Object.keys(store.servers)
}

export async function exportConfig(passphrase: string): Promise<Buffer> {
  const store = await loadStore()
  const key = await derivePassphraseKey(passphrase)
  return encryptData(key, JSON.stringify(store))
}

export async function importConfig(passphrase: string, data: Buffer): Promise<void> {
  let json: string
  try {
    const key = await derivePassphraseKey(passphrase, PBKDF2_ITERATIONS)
    json = await decryptData(key, data)
  } catch (err) {
    try {
      const legacyKey = await derivePassphraseKey(passphrase, LEGACY_PBKDF2_ITERATIONS)
      json = await decryptData(legacyKey, data)
    } catch {
      throw err
    }
  }
  const imported = JSON.parse(json) as ConfigStore

  const store = await loadStore()
  // Merge imported servers into local config
  for (const [name, config] of Object.entries(imported.servers)) {
    store.servers[name] = config
  }
  await saveStore(store)
}

File: packages/core-py/src/mcp_relay_core/storage/config_file.py
"""Encrypted config file management.

Reads/writes ~/.config/mcp/config.enc with the same format as the TS version.
"""

import json
import os
import threading
import time
from pathlib import Path
from typing import Any

from platformdirs import user_config_dir

from mcp_relay_core.storage.encryption import (
    LEGACY_PBKDF2_ITERATIONS,
    PBKDF2_ITERATIONS,
    decrypt_data,
    derive_file_key,
    derive_passphrase_key,
    encrypt_data,
)
from mcp_relay_core.storage.machine_id import get_machine_id, get_username

_DEFAULT_CONFIG_PATH = Path(user_config_dir("mcp", appauthor=False)) / "config.enc"
_MAX_RETRIES = 3
_BASE_DELAY_S = 0.1

# Allow overriding config path for testing
_config_path_override: str | None = None


def set_config_path(path: str | None) -> None:
    """Override config file path (for testing). Pass None to reset."""
    global _config_path_override
    _config_path_override = path


def _get_config_path() -> Path:
    if _config_path_override is not None:
        return Path(_config_path_override)
    return _DEFAULT_CONFIG_PATH


def _get_key() -> bytes:
    machine_id = get_machine_id()
    username = get_username()
    return derive_file_key(machine_id, username)


def _with_retry(fn: Any) -> Any:
    """Retry function on file busy errors."""
    for attempt in range(_MAX_RETRIES):
        try:
            return fn()
        except OSError as err:
            is_busy = getattr(err, "errno", None) in (11, 16, 35)  # EAGAIN, EBUSY, etc
            if not is_busy or attempt == _MAX_RETRIES - 1:
                raise
            time.sleep(_BASE_DELAY_S * (2**attempt))
    msg = "Unreachable"
    raise RuntimeError(msg)


def _load_store() -> dict[str, Any]:
    config_path = _get_config_path()
    if not config_path.exists():
        return {"version": 1, "servers": {}}

    machine_id = get_machine_id()
    username = get_username()
    data = config_path.read_bytes()

    try:
        key = derive_file_key(machine_id, username, PBKDF2_ITERATIONS)
        json_str = decrypt_data(key, data)
        return json.loads(json_str)
    except Exception as err:
        try:
            legacy_key = derive_file_key(machine_id, username, LEGACY_PBKDF2_ITERATIONS)
            json_str = decrypt_data(legacy_key, data)
            store = json.loads(json_str)
            # Auto-migrate to current iterations
            _save_store(store)
            return store
        except Exception:
            raise err from None


def _save_store(store: dict[str, Any]) -> None:
    config_path = _get_config_path()
    config_path.parent.mkdir(parents=True, exist_ok=True)
    key = _get_key()
    encrypted = encrypt_data(key, json.dumps(store))

    def _write() -> None:
        config_path.write_bytes(encrypted)

    _with_retry(_write)


def _schedule_restart() -> None:
    """Schedule a process exit to reload config via MCP client."""
    if "PYTEST_CURRENT_TEST" not in os.environ and "MCP_NO_RELOAD" not in os.environ:

        def _exit() -> None:
            time.sleep(1.0)
            os._exit(0)

        threading.Thread(target=_exit, daemon=True).start()


def read_config(server_name: str) -> dict[str, str] | None:
    """Read config for a server.

    Args:
        server_name: Server identifier.

    Returns:
        Config dict or None if not found.
    """
    store = _load_store()
    return store["servers"].get(server_name)


def write_config(server_name: str, config: dict[str, str]) -> None:
    """Write config for a server (merges with existing servers).

    Args:
        server_name: Server identifier.
        config: Key-value config dict.
    """
    store = _load_store()
    store["servers"][server_name] = config
    _save_store(store)
    _schedule_restart()


def delete_config(server_name: str) -> None:
    """Delete config for a server.

    Removes the config file entirely if no servers remain.

    Args:
        server_name: Server identifier.
    """
    store = _load_store()
    store["servers"].pop(server_name, None)

    config_path = _get_config_path()
    if not store["servers"]:
        if config_path.exists():
            config_path.unlink()
    else:
        _save_store(store)
    _schedule_restart()


def list_configs() -> list[str]:
    """List all configured server names.

    Returns:
        List of server name strings.
    """
    store = _load_store()
    return list(store["servers"].keys())


def export_config(passphrase: str) -> bytes:
    """Export all configs encrypted with a passphrase.

    Args:
        passphrase: Passphrase to encrypt export data.

    Returns:
        Encrypted bytes.
    """
    store = _load_store()
    key = derive_passphrase_key(passphrase)
    return encrypt_data(key, json.dumps(store))


def import_config(passphrase: str, data: bytes) -> None:
    """Import configs from encrypted export data, merging into local config.

    Args:
        passphrase: Passphrase to decrypt import data.
        data: Encrypted bytes from export_config.

    Raises:
        cryptography.exceptions.InvalidTag: If passphrase is wrong.
    """
    try:
        key = derive_passphrase_key(passphrase, PBKDF2_ITERATIONS)
        json_str = decrypt_data(key, data)
    except Exception as err:
        try:
            legacy_key = derive_passphrase_key(passphrase, LEGACY_PBKDF2_ITERATIONS)
            json_str = decrypt_data(legacy_key, data)
        except Exception:
            raise err from None
    imported = json.loads(json_str)

    store = _load_store()
    for name, config in imported["servers"].items():
        store["servers"][name] = config
    _save_store(store)
    _schedule_restart()
